ksp_plugin/pile_up.cpp

```diff
@@ -416,6 +404,55 @@ void PileUp::AppendToPart(DiscreteTrajectory<Barycentric>::Iterator it) const {
   }
 }
 
+DegreesOfFreedom<Barycentric> PileUp::RecomputeFromParts(
+    std::list<not_null<Part*>> const& parts) {
+  // First compute the overall mass and centre of mass of the pile-up.  Also
+  // compute the overall force applied to it.
+  // TODO(phl): We assume that forces are applied at the centre of mass of the
+  // pile-up, but they are really applied at the centre of mass of the parts, so
+  // this introduces a torque.
+  Mass pile_up_mass;
+  Vector<Force, Barycentric> pile_up_intrinsic_force;
+  BarycentreCalculator<DegreesOfFreedom<Barycentric>, Mass> calculator;
+  for (not_null<Part*> const part : parts) {
+    pile_up_intrinsic_force += part->intrinsic_force();
+    calculator.Add(part->degrees_of_freedom(), part->inertia_tensor().mass());
+  }
+  pile_up_mass = calculator.weight();
+  DegreesOfFreedom<Barycentric> const pile_up_barycentre = calculator.Get();
+
+  // Then compute the inertia tensor and the angular momentum of the pile-up.
+  InertiaTensor<RigidPileUp> pile_up_inertia_tensor;
+  Bivector<AngularMomentum, RigidPileUp> pile_up_angular_momentum;
+  RigidTransformation<RigidPileUp, Barycentric> const pile_up_to_barycentric(
+      RigidPileUp::origin,
+      pile_up_barycentre.position(),
+      Identity<RigidPileUp, Barycentric>().Forget());
+  for (not_null<Part*> const part : parts) {
+    RigidMotion<RigidPart, Barycentric> const& part_rigid_motion =
+        part->rigid_motion();
+    RigidTransformation<RigidPart, Barycentric> const& part_to_barycentric =
+        part_rigid_motion.rigid_transformation();
+    RigidTransformation<RigidPart, RigidPileUp> const part_to_pile_up =
+        pile_up_to_barycentric.Inverse() * part_to_barycentric;
+
+    InertiaTensor<RigidPart> const& part_inertia_tensor =
+        part->inertia_tensor();
+    pile_up_inertia_tensor += part->inertia_tensor().Transform(part_to_pile_up);
+
+    Bivector<AngularMomentum, RigidPart> const part_angular_momentum =
+        Anticommutator(part_inertia_tensor,
+                       part_rigid_motion.angular_velocity_of_to_frame());
+    pile_up_angular_momentum +=
+        part_to_pile_up.linear_map()(part_angular_momentum);
```

Comment:
This is incorrect; you are computing the angular momentum of the parts with respect to their centres of mass (in `RigidPart`) and then changing coordinates to `RigidPileUp`; this means that you are missing the contribution of the linear motion of the part to the angular momentum in `RigidPileUp`.

Reply:
Done (I think).